### ISSUE TITLE
How do I create task specific environments that are listed in 1M and 1B benchmarks?

### ISSUE THREAD

**Issue by NishanthVAnand** [MichaelTMatthews/Craftax#37]
I'd like to train RL agents on various tasks listed in craftax 1M and 1B benchmark (eg. eat-cow): https://arxiv.org/pdf/2402.16801, but I couldn't find the environment registration for these tasks. How do I create an environment corresponding to these tasks/benchmarks?

**Comment by MichaelTMatthews** [comment]
Hi Nishanth, I'm assuming you're saying that you want an environment in which reward is only received for a particular task (e.g. `EAT_COW`).  This isn't supported currently, but would be possible to do.

If you want to do this statically you could change the `achievement_mapping` function in `constants.py` to
```python
def achievement_mapping(achievement_value):
    if achievement_value == Achievement.EAT_COW.value:
        return 1
    else:
        return 0
```

If you want to dynamically change the target task then this is a bit harder.  You would need to add the target task to `EnvState` and use that in the reward calculation in `craftax_step`. You could change line 3081 in `game_logic.py` as follows from

```python
achievement_coefficients = ACHIEVEMENT_REWARD_MAP
```

to

```python
achievement_coefficients = jnp.zeros(len(Achievement)).at[state.target_achievement_index].set(1.)
```

where you would add `target_achievement_index` to `EnvState` and initialise it however you want.

Let me know if this helps and if you have any further questions.

**Comment by NishanthVAnand** [comment]
Thanks! This helps. I am indeed looking to change rewards for tasks over time. I will get back to you in case of any questions.

**Comment by NishanthVAnand** [comment]
I am wondering if the simple wrapper like this would do the trick instead of changing the constants file and the game logic. Any gotchas here?

```python
class RewardWrapper(GymnaxWrapper):
    """Log the episode returns and lengths."""

    def __init__(self, env, achievement):
        super().__init__(env)
        self.achievement = Achievement[achievement].value

    @partial(jax.jit, static_argnums=(0, 4))
    def step(
        self,
        key: chex.PRNGKey,
        state,
        action: Union[int, float],
        params=None,
    ):
        obs, env_state, _, done_old, info = self._env.step(key, state, action, params)
        achievement_done = env_state.achievements[self.achievement]
        achievement_reward = achievement_done.astype(jnp.float32)
        init_health = state.player_health
        health_reward = (env_state.player_health - init_health) * 0.1  # health reward original
        reward = achievement_reward + health_reward
        done = jnp.logical_or(done_old, achievement_done)
        return obs, env_state, reward, done, info
```

**Comment by MichaelTMatthews** [comment]
Hi Nishanth,

Good spot, this is probably an easier way to do it - code looks good to me.

**Comment by NishanthVAnand** [comment]
Thanks, Michael!

Although, done flag is creating an issue that I am unable to understand. I want to terminate the episode when the agent solves the specified achievement. This is happening with the current code, but the final reward is being ignored by the agent.

**Comment by NishanthVAnand** [comment]
Perhaps, [this line](https://github.com/MichaelTMatthews/Craftax_Baselines/blob/a2a029d713af643edeb0e101fa1844f30c65930f/wrappers.py#L116) is the culprit? Instead of taking a `step` by calling the parent class method, it is overriding it.
`obs_st, state_st, reward, done, info = self.step_fn(rngs, state, action, params)`

**Comment by MichaelTMatthews** [comment]
Which order are you applying the wrappers? Applying your RewardWrapper first and then the optimistic reset wrapper should produce the desired behaviour

**Comment by MichaelTMatthews** [comment]
Closing due to inactivity, feel free to re-open if you'd like to pursue this further
